// nodemailer-mjml/dist/helpers/buildMjmlTemplate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildMjmlTemplate = void 0;
const promises_1 = require("fs/promises");
const mjml_1 = __importDefault(require("mjml"));
const mustache_1 = require("mustache");
const path_1 = require("path");
const defaultPluginOptions_1 = require("../constants/defaultPluginOptions");
const checkMjmlError_1 = require("../helpers/checkMjmlError");
const buildLayout_1 = require("./buildLayout");
const minifyTemplate_1 = require("./minifyTemplate");
const buildMjmlTemplate = async (options, sendMailTemplateOptions) => {
    const renderOptions = {
        ...defaultPluginOptions_1.defaultPluginOptions,
        ...options,
        mjmlOptions: {
            ...defaultPluginOptions_1.defaultPluginOptions.mjmlOptions,
            ...options.mjmlOptions
        }
    };
    const { templateData, templateName, templateLayoutName, templateLayoutSlots } = sendMailTemplateOptions;
    const mjmlTemplatePath = (0, path_1.join)(options.templateFolder, `${templateLayoutName ? templateLayoutName : templateName}.mjml`);
    const rawMjmlTemplate = await (async () => {
        if (templateLayoutName) {
            return (0, buildLayout_1.buildLayout)({
                templateFolder: options.templateFolder,
                templateLayoutName,
                templateLayoutSlots: templateLayoutSlots ?? {},
                templatePartialsFolder: options.templatePartialsFolder
            }, templateData);
        }
        try {
            return await (0, promises_1.readFile)(mjmlTemplatePath, "utf-8");
        }
        catch {
            throw new Error(`[nodemailer-mjml] - Could not read mjml template at path: ${mjmlTemplatePath}`);
        }
    })();
    const mjmlOutput = (0, mjml_1.default)(rawMjmlTemplate, {
        filePath: mjmlTemplatePath,
        ...renderOptions.mjmlOptions
    });
    (0, checkMjmlError_1.checkMjmlError)(mjmlOutput);
    const finalHtmlOutput = renderOptions.minifyHtmlOutput ? await (0, minifyTemplate_1.minifyTemplate)(mjmlOutput.html, renderOptions.templateMinifierOptions) : mjmlOutput.html;
    const shouldRunMustacheCompiler = !!templateData && Object.keys(templateData ?? {}).length > 0;
    return shouldRunMustacheCompiler ? (0, mustache_1.render)(finalHtmlOutput, templateData) : finalHtmlOutput;
};
exports.buildMjmlTemplate = buildMjmlTemplate;
